plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergPartitionEvolutionOnSameColumn.java

```diff
@@ -0,0 +1,167 @@
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("SELECT CAST(ROW('In') AS ROW(part_trunc varchar))");
+
+            assertUpdate("ALTER TABLE " + table.getName() + " SET PROPERTIES partitioning = ARRAY['truncate(part, 4)']");
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (2, 'Poland')", 1);
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'India'), (2, VARCHAR 'Poland')");
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("VALUES " +
+                            "ROW(CAST(ROW('In', null) AS ROW(part_trunc varchar, part_trunc_2 varchar))), " +
+                            "ROW(CAST(ROW(null, 'Pola') AS ROW(part_trunc varchar, part_trunc_2 varchar)))");
+        }
+    }
+
+    @Test
+    void testFilesPartitionEvolutionUsingBucketOnSameColumn()
+    {
+        try (TestTable table = newTrinoTable("test_files_table", "(id int, part varchar) WITH (partitioning = ARRAY['bucket(part, 2)'])")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (1, 'India')", 1);
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'India')");
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("SELECT CAST(ROW(0) AS ROW(part_bucket int))");
+
+            assertUpdate("ALTER TABLE " + table.getName() + " SET PROPERTIES partitioning = ARRAY['bucket(part, 4)']");
+
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (2, 'Poland')", 1);
+            assertThat(query("SELECT * FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'India'), (2, VARCHAR 'Poland')");
+            assertThat(query("SELECT partition FROM \"" + table.getName() + "$files\""))
+                    .matches("VALUES " +
+                            "ROW(CAST(ROW(0, null) AS ROW(part_bucket int, part_bucket_2 int))), " +
+                            "ROW(CAST(ROW(null, 1) AS ROW(part_bucket int, part_bucket_2 int)))");
+        }
+    }
+
+    @Test // https://github.com/trinodb/trino/issues/26109
+    void testFilesPartitionEvolutionWithTruncateConflictFix()
+    {
+        String tableName = "test_iceberg_partition_evolution_" + randomNameSuffix();
+
+        minio.copyResources("iceberg/conflict_truncate", BUCKET_NAME, "conflict_truncate");
+        assertUpdate(format(
+                "CALL system.register_table(CURRENT_SCHEMA, '%s', '%s')",
+                tableName,
+                format("s3://%s/conflict_truncate", BUCKET_NAME)));
+
+        assertThat(query("SELECT * FROM " + tableName))
+                .matches("VALUES (VARCHAR 'abc'), (VARCHAR 'abcd')");
+
+        assertThatThrownBy(() -> computeActual("SELECT partition FROM \"" + tableName + "$files\""))
```

Comment:
This is not straightforward to understand why is it happening.


Taken from `00003-6f8f2126-abe7-449f-93dd-c10f484de0dd.metadata.json‎`:

```
"partition-specs":[{"spec-id":0,"fields":[{"name":"a_trunc","transform":"truncate[1]","source-id":1,"field-id":1000}]},{"spec-id":1,"fields":[{"name":"a_trunc","transform":"truncate[10]","source-id":1,"field-id":1001}]}]
```

Please add a comment explaining that the current state of the table is corrupted and why it is corrupted.